File: src-tauri/src/edit/remap_doc.rs
use crate::edit::captions::{Caption, CaptionWord};
use crate::edit::model::{EditDoc, Trim};
use crate::export::remap::TimeMap;

fn span(map: &TimeMap, start_ms: u32, end_ms: u32) -> Option<(u32, u32)> {
    let (s, e) = (map.out_of(start_ms), map.out_of(end_ms));
    (e > s).then_some((s, e))
}

pub fn remap_doc(doc: &EditDoc, map: &TimeMap) -> EditDoc {
    let mut out = doc.clone();
    out.zooms = doc
        .zooms
        .iter()
        .filter_map(|z| {
            span(map, z.start_ms, z.end_ms).map(|(s, e)| {
                let mut z = z.clone();
                z.start_ms = s;
                z.end_ms = e;
                z
            })
        })
        .collect();
    out.layout = doc
        .layout
        .iter()
        .filter_map(|l| {
            span(map, l.start_ms, l.end_ms).map(|(s, e)| {
                let mut l = l.clone();
                l.start_ms = s;
                l.end_ms = e;
                l
            })
        })
        .collect();
    out.effects = doc
        .effects
        .iter()
        .filter_map(|f| {
            span(map, f.start_ms, f.end_ms).map(|(s, e)| {
                let mut f = f.clone();
                f.start_ms = s;
                f.end_ms = e;
                f
            })
        })
        .collect();
    out.captions = doc
        .captions
        .iter()
        .filter_map(|c| {
            span(map, c.start_ms, c.end_ms).map(|(s, e)| {
                let words = c
                    .words
                    .iter()
                    .filter_map(|w| {
                        span(map, w.start_ms, w.end_ms).map(|(ws, we)| CaptionWord {
                            start_ms: ws,
                            end_ms: we,
                            text: w.text.clone(),
                        })
                    })
                    .collect();
                Caption {
                    start_ms: s,
                    end_ms: e,
                    words,
                    ..c.clone()
                }
            })
        })
        .collect();
    out.camera_moves = doc
        .camera_moves
        .iter()
        .map(|m| {
            let mut m = m.clone();
            m.t_ms = map.out_of(m.t_ms);
            m
        })
        .collect();
    out.trim = Trim::default();
    out.cuts.clear();
    out.speed.clear();
    out.clip_ms = map.out_dur_ms();
    out
}

#[cfg(test)]
#[path = "remap_doc_tests.rs"]
mod tests;

File: src/shared/math/remapDoc.ts
import type { EditDoc } from "../edit";
import { outDurMs, outOf, type TimeMap } from "./remap";

export function remapDoc(doc: EditDoc, map: TimeMap): EditDoc {
  const span = (s: number, e: number): [number, number] | null => {
    const [a, b] = [outOf(map, s), outOf(map, e)];
    return b > a ? [a, b] : null;
  };
  const moved = <T extends { start_ms: number; end_ms: number }>(list: T[]): T[] =>
    list.flatMap((r) => {
      const sp = span(r.start_ms, r.end_ms);
      return sp ? [{ ...r, start_ms: sp[0], end_ms: sp[1] }] : [];
    });
  return {
    ...doc,
    zooms: moved(doc.zooms),
    layout: moved(doc.layout),
    effects: moved(doc.effects),
    texts: moved(doc.texts),
    captions: doc.captions.flatMap((c) => {
      const sp = span(c.start_ms, c.end_ms);
      return sp ? [{ ...c, start_ms: sp[0], end_ms: sp[1], words: moved(c.words) }] : [];
    }),
    camera_moves: doc.camera_moves.map((m) => ({ ...m, t_ms: outOf(map, m.t_ms) })),
    trim: { in_ms: 0, out_ms: 0 },
    cuts: [],
    speed: [],
    clips: [],
    clip_ms: outDurMs(map),
  };
}
